File: backend-express/src/routes/analytics.js
const express = require('express');
const { auth } = require('../middleware/auth');
const { cacheGet, cacheSet } = require('../db');
const {
  Employee, Candidate, Interview, Attendance, Payroll, Performance, Leave, Job, JobApplication,
} = require('../models');
const { ensureCandidateForUser, buildInterviewFilterForUser } = require('../lib/candidateLink');
const { dedupeInterviewsByRole } = require('../lib/interviewDedup');

const router = express.Router();
const CACHE_KEY = 'analytics:dashboard:v3';

function isInterviewPastDeadline(item) {
  const d = item.deadlineAt || item.scheduledAt;
  if (!d) return false;
  return Date.now() > new Date(d).getTime();
}

const MONTHS = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

router.get('/dashboard', auth(['management_admin', 'senior_manager', 'hr_recruiter']), async (req, res) => {
  const cached = await cacheGet(CACHE_KEY);
  if (cached) return res.json(JSON.parse(cached));

  const today = new Date().toISOString().split('T')[0];
  const sixMonthsAgo = new Date();
  sixMonthsAgo.setMonth(sixMonthsAgo.getMonth() - 6);

  const [
    totalEmployees,
    totalApplications,
    openJobs,
    scheduledInterviews,
    completedInterviews,
    attendanceToday,
    payrolls,
    avgJdScore,
    monthlyApps,
    appStatusAgg,
  ] = await Promise.all([
    Employee.countDocuments({ status: 'active' }),
    JobApplication.countDocuments(),
    Job.countDocuments({ status: { $ne: 'closed' } }),
    Interview.find({ status: 'scheduled' }).lean(),
    Interview.find({ status: 'completed' }).lean(),
    Attendance.countDocuments({ date: today }),
    Payroll.find().sort({ month: -1 }).limit(12).lean(),
    JobApplication.aggregate([
      { $match: { jdScore: { $gt: 0 } } },
      { $group: { _id: null, avg: { $avg: '$jdScore' } } },
    ]),
    JobApplication.aggregate([
      { $match: { appliedAt: { $gte: sixMonthsAgo } } },
      { $group: { _id: { $month: '$appliedAt' }, count: { $sum: 1 } } },
      { $sort: { _id: 1 } },
    ]),
    JobApplication.aggregate([
      { $group: { _id: '$status', count: { $sum: 1 } } },
    ]),
  ]);

  const uniqueScheduled = dedupeInterviewsByRole(scheduledInterviews, isInterviewPastDeadline)
    .filter((i) => i.status === 'scheduled' && !isInterviewPastDeadline(i));

  const avgScore = Math.round((avgJdScore[0]?.avg || 0) * 100) / 100;

  const statusCounts = Object.fromEntries(
    ['applied', 'shortlisted', 'rejected', 'interview_scheduled'].map((s) => [s, 0])
  );
  for (const row of appStatusAgg) {
    if (row._id) statusCounts[row._id] = row.count;
  }

  const monthly = monthlyApps.map((m) => ({
    month: MONTHS[m._id - 1] || `M${m._id}`,
    count: m.count,
  }));

  const skillAgg = await JobApplication.aggregate([
    { $unwind: '$skills' },
    { $group: { _id: '$skills', count: { $sum: 1 } } },
    { $sort: { count: -1 } },
    { $limit: 15 },
  ]);
  const skillTrends = skillAgg.map((s) => ({ skill: s._id, count: s.count }));

  const shortlisted = statusCounts.shortlisted || 0;
  const rejected = statusCounts.rejected || 0;

  const result = {
    total_employees: totalEmployees,
    total_applications: totalApplications,
    selected_candidates: shortlisted,
    rejected_candidates: rejected,
    average_ai_score: avgScore,
    attendance_today: attendanceToday,
    open_jobs: openJobs,
    scheduled_interviews: uniqueScheduled.length,
    hiring_funnel: statusCounts,
    skill_trends: skillTrends,
    interview_performance: {
      avg_technical: completedInterviews.reduce((s, i) => s + (i.technicalScore || 0), 0) / Math.max(completedInterviews.length, 1),
      avg_communication: completedInterviews.reduce((s, i) => s + (i.communicationScore || 0), 0) / Math.max(completedInterviews.length, 1),
      total_interviews: completedInterviews.length,
    },
    predictions: {
      hiring_success_probability: Math.min(95, Math.round((shortlisted / Math.max(totalApplications, 1)) * 100 + 10)),
      employee_growth_rate: totalEmployees > 0 ? Math.round((totalEmployees / Math.max(totalEmployees - 1, 1)) * 100 - 100) : 0,
      attrition_risk_avg: 22,
    },
    charts: {
      funnel_bar: Object.entries(statusCounts).map(([name, value]) => [name.replace(/_/g, ' '), value]),
      monthly_applications: monthly,
      salary_analytics: payrolls.map((p) => ({ month: p.month, netPay: p.netPay })),
      skill_heatmap: skillTrends.slice(0, 10),
    },
  };

  await cacheSet(CACHE_KEY, result, 60);
  res.json(result);
});

router.get('/candidate-portal', auth(['candidate']), async (req, res) => {
  const candidate = await ensureCandidateForUser(req.user);
  const openJobs = await Job.countDocuments({ status: { $ne: 'closed' } });

  const interviewFilter = await buildInterviewFilterForUser(req.user);
  const [applicationsCount, applications, interviews] = await Promise.all([
    JobApplication.countDocuments({
      $or: [{ candidateId: candidate.id }, { userId: req.user.id }],
    }),
    JobApplication.find({
      $or: [{ candidateId: candidate.id }, { userId: req.user.id }],
    }).sort({ appliedAt: -1 }).limit(10).lean(),
    Interview.find(interviewFilter)
      .sort({ scheduledAt: -1, createdAt: -1 }).lean(),
  ]);

  const uniqueInterviews = dedupeInterviewsByRole(interviews, isInterviewPastDeadline);
  const upcoming = uniqueInterviews.filter(
    (i) => i.status === 'scheduled' && !isInterviewPastDeadline(i),
  );

  res.json({
    applications_count: applicationsCount,
    open_jobs: openJobs,
    interviews_scheduled: upcoming.length,
    interviews_completed: uniqueInterviews.filter((i) => i.status === 'completed').length,
    recent_applications: applications,
    upcoming_interviews: upcoming,
    candidate_status: candidate.status,
  });
});

router.get('/portal', auth(['employee']), async (req, res) => {
  const emp = await Employee.findOne({ 'personalDetails.email': req.user.email }).lean();
  const [attendance, payroll, performance, leaves] = await Promise.all([
    Attendance.find({ employeeId: emp?.id }).sort({ date: -1 }).limit(10).lean(),
    Payroll.find({ employeeId: emp?.id }).sort({ month: -1 }).limit(3).lean(),
    Performance.findOne({ employeeId: emp?.id }).sort({ createdAt: -1 }).lean(),
    Leave.find({ employeeId: emp?.id }).sort({ createdAt: -1 }).limit(5).lean(),
  ]);
  res.json({ employee: emp, attendance, payroll, performance, leaves });
});

module.exports = router;

File: backend/app/api/routes/analytics.py
from datetime import datetime, timezone, timedelta
from fastapi import APIRouter, Depends
from motor.motor_asyncio import AsyncIOMotorDatabase
from app.database import get_db
from app.models.enums import UserRole, CandidateStatus
from app.schemas.analytics import AnalyticsResponse
from app.core.auth import require_roles

router = APIRouter(prefix="/analytics", tags=["Analytics"])

MONTH_NAMES = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]


async def _monthly_applications(database) -> list[dict]:
    six_months_ago = datetime.now(timezone.utc) - timedelta(days=180)
    pipeline = [
        {"$match": {"created_at": {"$gte": six_months_ago}}},
        {"$group": {
            "_id": {"$month": "$created_at"},
            "count": {"$sum": 1},
            "year": {"$first": {"$year": "$created_at"}},
        }},
        {"$sort": {"_id": 1}},
    ]
    results = await database.candidates.aggregate(pipeline).to_list(12)
    month_map = {r["_id"]: r["count"] for r in results}

    now = datetime.now(timezone.utc)
    monthly = []
    for i in range(5, -1, -1):
        d = now - timedelta(days=30 * i)
        month_num = d.month
        monthly.append({
            "month": MONTH_NAMES[month_num - 1],
            "year": d.year,
            "count": month_map.get(month_num, 0),
        })
    return monthly


async def _time_to_hire_days(database) -> float:
    selected = await database.candidates.find(
        {"status": {"$in": [CandidateStatus.SELECTED.value, CandidateStatus.ONBOARDING.value]}},
        {"_id": 0, "created_at": 1, "updated_at": 1},
    ).to_list(500)

    if not selected:
        return 0.0

    total_days = 0
    for c in selected:
        created = c.get("created_at")
        updated = c.get("updated_at", created)
        if created and updated:
            delta = updated - created if isinstance(updated, datetime) else timedelta(days=14)
            total_days += delta.days if hasattr(delta, "days") else 14

    return round(total_days / len(selected), 1)


@router.get("/dashboard", response_model=AnalyticsResponse)
async def get_dashboard(
    database: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(require_roles(UserRole.ADMIN, UserRole.RECRUITER)),
):
    total = await database.candidates.count_documents({})
    selected = await database.candidates.count_documents({
        "status": {"$in": [CandidateStatus.SELECTED.value, CandidateStatus.ONBOARDING.value]}
    })
    rejected = await database.candidates.count_documents({"status": CandidateStatus.REJECTED.value})

    agg = await database.candidates.aggregate([
        {"$group": {"_id": None, "avg": {"$avg": "$ai_score"}}}
    ]).to_list(1)
    avg_score = agg[0]["avg"] if agg else 0

    cursor = database.candidates.find({}, {"_id": 0, "skills": 1, "ai_score": 1})
    candidates = await cursor.to_list(length=10000)

    skill_counts: dict[str, int] = {}
    for c in candidates:
        for skill in c.get("skills") or []:
            skill_counts[skill] = skill_counts.get(skill, 0) + 1
    skill_trends = sorted(
        [{"skill": k, "count": v} for k, v in skill_counts.items()],
        key=lambda x: x["count"],
        reverse=True,
    )[:15]

    interviews = await database.interviews.find(
        {"status": "completed"}, {"_id": 0}
    ).to_list(1000)

    interview_perf = {
        "avg_technical": sum(i.get("technical_score", 0) for i in interviews) / max(len(interviews), 1),
        "avg_communication": sum(i.get("communication_score", 0) for i in interviews) / max(len(interviews), 1),
        "avg_confidence": sum(i.get("confidence_score", 0) for i in interviews) / max(len(interviews), 1),
        "total_interviews": len(interviews),
    }

    status_counts = {}
    for status in CandidateStatus:
        status_counts[status.value] = await database.candidates.count_documents({"status": status.value})

    time_to_hire = await _time_to_hire_days(database)
    monthly_apps = await _monthly_applications(database)

    hire_rate = (selected / max(total, 1)) * 100
    interview_rate = len(interviews) / max(total, 1) * 100

    return AnalyticsResponse(
        hiring_funnel=status_counts,
        total_applications=total,
        selected_candidates=selected,
        rejected_candidates=rejected,
        average_ai_score=round(float(avg_score or 0), 2),
        skill_trends=skill_trends,
        interview_performance=interview_perf,
        predictions={
            "hiring_success_probability": round(min(95, hire_rate * 0.6 + interview_rate * 0.4 + 10), 1),
            "time_to_hire_days": time_to_hire or 28,
            "quality_of_hire_score": round(float(avg_score or 0) * 0.85 + hire_rate * 0.15, 1),
        },
        charts={
            "funnel_bar": list(status_counts.items()),
            "score_distribution": [
                {"range": "90-100", "count": sum(1 for c in candidates if c.get("ai_score", 0) >= 90)},
                {"range": "70-89", "count": sum(1 for c in candidates if 70 <= c.get("ai_score", 0) < 90)},
                {"range": "50-69", "count": sum(1 for c in candidates if 50 <= c.get("ai_score", 0) < 70)},
                {"range": "0-49", "count": sum(1 for c in candidates if c.get("ai_score", 0) < 50)},
            ],
            "monthly_applications": monthly_apps,
            "skill_heatmap": skill_trends[:10],
        },
    )
